xds/src/main/java/io/grpc/xds/internal/rbac/engine/AuthorizationEngine.java

```diff
@@ -0,0 +1,203 @@
+    private final Action action;
+    private final ImmutableMap<String, Expr> conditions;
+
+    public RbacEngine(Action action, ImmutableMap<String, Expr> conditions) {
+      this.action = action;
+      this.conditions = conditions;
+    }
+  }
+
+  private final List<RbacEngine> rbacEngines;
+
+  /**
+   * Creates a CEL-based Authorization Engine from a list of Envoy RBACs.
+   * The constructor can take either one or two Envoy RBACs.
+   * When it takes two RBACs, the order has to be a RBAC with DENY action
+   * followed by a RBAC with ALLOW action.
+   * @param rbacPolicies input Envoy RBAC list.
+   * @throws IllegalArgumentException if the user inputs an invalid RBAC list.
+   */
+  public AuthorizationEngine(ImmutableList<RBAC> rbacPolicies) throws IllegalArgumentException {
+    if (rbacPolicies.size() < 1 || rbacPolicies.size() > 2) {
+      throw new IllegalArgumentException(
+        "Invalid RBAC list size, must provide either one RBAC or two RBACs. ");
+    } 
+    if (rbacPolicies.size() == 2 && (rbacPolicies.get(0).getAction() != Action.DENY 
+        || rbacPolicies.get(1).getAction() != Action.ALLOW)) {
+      throw new IllegalArgumentException( "Invalid RBAC list, " 
+          + "must provide a RBAC with DENY action followed by a RBAC with ALLOW action. ");
+    }
+    this.rbacEngines = new ArrayList<>();
+    for (RBAC rbac : rbacPolicies) {
+      Map<String, Expr> conditions = new HashMap<>();
+      for (Map.Entry<String, Policy> rbacPolicy: rbac.getPolicies().entrySet()) {
+        conditions.put(rbacPolicy.getKey(), rbacPolicy.getValue().getCondition());
+      }
+      this.rbacEngines.add(new RbacEngine(
+          Preconditions.checkNotNull(rbac.getAction()), 
+          Preconditions.checkNotNull(ImmutableMap.copyOf(conditions))));
+    }
+  }
+
+  /**
+   * The evaluate function performs the core authorization mechanism of CEL Evaluation Engine.
+   * Determines whether a gRPC call is allowed, denied, or unable to decide.
+   * @param args evaluate argument that is used to evaluate the RBAC conditions.
+   * @return an AuthorizationDecision generated by CEL Evaluation Engine.
+   * @throws InterpreterException if something goes wrong in CEL library.
+   */
+  public AuthorizationDecision evaluate(EvaluateArgs<ReqT, RespT> args) 
+      throws IllegalArgumentException {
+    if (this.rbacEngines.size() < 1 || this.rbacEngines.size() > 2) {
+      throw new IllegalArgumentException(
+        "Invalid RBAC list size, must provide either one RBAC or two RBACs. ");
+    } 
+    if (this.rbacEngines.size() == 2 && (this.rbacEngines.get(0).action != Action.DENY 
+        || this.rbacEngines.get(1).action != Action.ALLOW)) {
+      throw new IllegalArgumentException( "Invalid RBAC list, " 
+          + "must provide a RBAC with DENY action followed by a RBAC with ALLOW action. ");
+    }
+    AuthorizationDecision.Decision authorizationDecision = null;
+    List<String> matchingPolicyNames = new ArrayList<>();
```

Comment:
Now when there is a match, we will only return the policy that first matches and return it in a list. So if there is a match, we will do an early stop and return a list containing only 1 item. We might also want to slightly change the logic we check the RBACs(So we will first check the first RABC, if it's undecided, then the result would be the decision of second RBAC). You can checkout the latest design details in @jiangtaoli2016 's doc, or check with Elizabeth. Please LMK if you have any questions. Thanks!

Reply:
Once you have allowEngine and denyEngine. See discussion above. The evaluation logic is as follows:

First iterate denyEngine's map (could be empty map). If there is match, immediately return deny. If there are unknown results, return undecided. If all non-match, we iterate allowEngine. 

Once we enter allowEngine, if there is a match, immediately return allow. In the end of iteration, if there are unknown rules, return undecided. If all non-match, return deny.

The above logic is the same regardless if the constructor takes 1 or 2 RBAC rules.